src/Vault.sol

```diff
@@ -0,0 +1,398 @@
+        return _convertToAssets(asset(), shares, Math.Rounding.Ceil);
+    }
+
+    function previewWithdraw(uint256 assets) public view returns (uint256) {
+        return _convertToShares(asset(), assets, Math.Rounding.Ceil);
+    }
+
+    function previewRedeem(uint256 shares) public view returns (uint256) {
+        return _convertToAssets(asset(), shares, Math.Rounding.Floor);
+    }
+
+    function deposit(uint256 assets, address receiver) public nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
+
+        uint256 shares = previewDeposit(assets);
+        _deposit(asset(), _msgSender(), receiver, assets, shares);
+
+        return shares;
+    }
+
+    function mint(uint256 shares, address receiver) public virtual nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
+
+        uint256 assets_ = previewMint(shares);
+        _deposit(asset(), _msgSender(), receiver, assets_, shares);
+
+        return assets_;
+    }
+
+    function withdraw(uint256 assets, address receiver, address owner) public nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
+
+        uint256 maxAssets = maxWithdraw(owner);
+        if (assets > maxAssets) {
+            revert ExceededMaxWithdraw(owner, assets, maxAssets);
+        }
+
+        uint256 shares = previewWithdraw(assets);
+
+        _withdraw(_msgSender(), receiver, owner, assets, shares);
+
+        return shares;
+    }
+
+    // QUESTION: How to handle this in v1 with async withdraws
+    function redeem(uint256 shares, address receiver, address owner) public nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
+
+        uint256 maxShares = maxRedeem(owner);
+        if (shares > maxShares) {
+            revert ExceededMaxRedeem(owner, shares, maxShares);
+        }
+
+        uint256 assets = previewRedeem(shares);
+
+        _withdraw(_msgSender(), receiver, owner, assets, shares);
+
+        return assets;
+    }
+
+    //// 4626-MAX ////
+
+    function getAssets() public view returns (address[] memory) {
+        return _getAssetStorage().list;
+    }
+
+    function getAsset(address asset_) public view returns (AssetParams memory) {
+        return _getAssetStorage().assets[asset_];
+    }
+
+    function getStrategies() public view returns (address[] memory) {
+        return _getStrategyStorage().list;
+    }
+
+    function getStrategy(address asset_) public view returns (StrategyParams memory) {
+        return _getStrategyStorage().strategies[asset_];
+    }
+
+    function previewDepositAsset(address asset_, uint256 assets_) public view returns (uint256) {
+        if (!_getAssetStorage().assets[asset_].active) revert InvalidAsset();
+        return _convertToShares(asset_, assets_, Math.Rounding.Floor);
+    }
+
+    function depositAsset(address asset_, uint256 assets_, address receiver) public nonReentrant returns (uint256) {
+        if (paused()) revert Paused();
```

Comment:
personal preference for 

if (paused()) {
revert Paused();
}

as it allows no room for error when adding a second line

Reply:
implemented this suggestion